# Задача 1
Скачать данные по ссылке https://www.kaggle.com/datasets/ionaskel/laptop-prices \
Считать данные с помощью pandas\
Вывести на экран первые 5 строк

In [1]:
from google.colab import files

uploaded = files.upload()

Saving kc-house-data.csv to kc-house-data.csv


In [2]:
import pandas as pd

In [5]:
df = pd.read_csv('kc-house-data.csv', encoding='cp1251', index_col='id')
df.head()

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
7129300520  20141013T000000  221900.0         3       1.00         1180   
6414100192  20141209T000000  538000.0         3       2.25         2570   
5631500400  20150225T000000  180000.0         2       1.00          770   
2487200875  20141209T000000  604000.0         4       3.00         1960   
1954400510  20150218T000000  510000.0         3       2.00         1680   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
7129300520      5650     1.0           0     0          3      7        1180   
6414100192      7242     2.0           0     0          3      7        2170   
5631500400     10000     1.0           0     0          3      6         770   
2487200875      5000     1.0           0     0          5      7        1050   
1954400510      8080     1.0           0     0          3      8        1680   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновации и годом постройки дома. Если реновации дома не было, то в новом признаке поставьте "0"

In [21]:
def renovation_time(ren_year, built_year):
    if ren_year:
        return ren_year - built_year
    else:
        return ren_year


df['delta_renovated'] = df.apply(lambda x: renovation_time(x['yr_renovated'], x['yr_built']), axis=1)
df.head(2)

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
7129300520  20141013T000000  221900.0         3       1.00         1180   
6414100192  20141209T000000  538000.0         3       2.25         2570   

            sqft_lot  floors  waterfront  view  condition  ...  sqft_above  \
id                                                         ...               
7129300520      5650     1.0           0     0          3  ...        1180   
6414100192      7242     2.0           0     0          3  ...        2170   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   

            sqft_living15  sqft_lot15  delta_renovated  
id                                                      
7129300520           1340        5650                0  
6414100192           1690        7639               40  

[2 rows x 21 columns]

1.3 Создайте признаки года продажи, месяца продажи

In [25]:
df['year_of_sale'] = df['date'].apply(lambda x: int(x[:4]))
df.sample()

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
8068000440  20141226T000000  399000.0         3       1.75         1620   

            sqft_lot  floors  waterfront  view  condition  ...  yr_built  \
id                                                         ...             
8068000440     10000     1.5           0     0          5  ...      1918   

            yr_renovated  zipcode      lat     long  sqft_living15  \
id                                                                   
8068000440             0    98178  47.5091 -122.262           1880   

            sqft_lot15  delta_renovated  year_of_sale  month_of_sale  
id                                                                    
8068000440       10000                0          2014             12  

[1 rows x 23 columns]

In [24]:
df['month_of_sale'] = df['date'].apply(lambda x: int(x[4:6]))
df.sample()

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
4358700141  20150427T000000  480000.0         4       1.75         1840   

            sqft_lot  floors  waterfront  view  condition  ...  yr_built  \
id                                                         ...             
4358700141      9250     1.0           0     0          4  ...      1956   

            yr_renovated  zipcode     lat     long  sqft_living15  sqft_lot15  \
id                                                                              
4358700141             0    98133  47.708 -122.337           1520        9250   

            delta_renovated  year_of_sale  month_of_sale  
id                                                        
4358700141                0          2015              4  

[1 rows x 23 columns]

1.4 Удалите признаки date, zipcode, lat, long

In [26]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.head(2)

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
7129300520  221900.0         3       1.00         1180      5650     1.0   
6414100192  538000.0         3       2.25         2570      7242     2.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
7129300520           0     0          3      7        1180              0   
6414100192           0     0          3      7        2170            400   

            yr_built  yr_renovated  sqft_living15  sqft_lot15  \
id                                                              
7129300520      1955             0           1340        5650   
6414100192      1951          1991           1690        7639   

            delta_renovated  year_of_sale  month_of_sale  
id                                                        
7129300520                0          2014             10  
6414100192               40          2014             12

# Задача 2
Создайте датафрейм с клиентами: clients\
Присоедините к таблице clients данные по домам\
Это нужно, чтобы понимать, какие дома покупались клиентами\
house_id - это индексы датафрейма с домами

In [35]:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 
              4213, 3113, 4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720, 9510920050, 
             6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375, 7920100025, 1951600150, 
             809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030, 3303980210, 
             7731100066, 8146100580, 825069097, 3889100029, 9524100196]
})

clients.head(2)

client_id    house_id
0       1459  8965450190
1       4684  6823100225

2.1 Присоедините к таблице clients данные по домам через метод join

In [36]:
clients_id = clients.set_index('house_id')
clients_id.head(2)

client_id
house_id             
8965450190       1459
6823100225       4684

In [37]:
clients_id.join(df).head()

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
339600110        4809  395000.0         3       2.50         1610      3755   
809001400        3113  925000.0         3       1.00         1630      3200   
825069097        1764  770000.0         3       2.50         2650     40705   
1099600250       2242  260000.0         3       1.75         1710      6400   
1189000207       2308  387000.0         2       2.50         1170      1394   

            floors  waterfront  view  condition  grade  sqft_above  \
339600110      2.0           0     0          3      7        1610   
809001400      1.5           0     0          4      8        1630   
825069097      2.0           0     0          3      9        2650   
1099600250     1.0           0     0          4      7        1240   
1189000207     2.0           0     0          3      8        1170   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
339600110               0      1987             0           1300        3823   
809001400               0      1912             0           1710        3600   
825069097               0      1994             0           2550       42625   
1099600250            470      1976             0           1600        6400   
1189000207              0      2001             0           1250        3136   

            delta_renovated  year_of_sale  month_of_sale  
339600110                 0          2014              9  
809001400                 0          2014              9  
825069097                 0          2014              6  
1099600250                0          2014             12  
1189000207                0          2014             10

2.2 Присоедините к таблице clients данные по домам через метод merge

In [40]:
merged_clients_df = clients.merge(df, left_on='house_id', right_on='id', how='left')
merged_clients_df.head(2)

client_id    house_id     price  bedrooms  bathrooms  sqft_living  \
0       1459  8965450190  295000.0         3       2.50         1500   
1       4684  6823100225  700000.0         4       1.75         1870   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      3060     2.0           0     0  ...      7        1500              0   
1      6000     1.0           0     0  ...      8        1670            200   

   yr_built  yr_renovated  sqft_living15  sqft_lot15  delta_renovated  \
0      1994             0           2700        7734                0   
1      1949             0           1710        6000                0   

   year_of_sale  month_of_sale  
0          2015              2  
1          2015              4  

[2 rows x 21 columns]

# Задача 3
Составьте несколько сводных таблиц

3.1 Найдите среднюю стоимость домов в зависимости от количества спален.
Отсортируйте от меньшей стоимости к большей

In [43]:
grouped_df = df.groupby('bedrooms')
grouped_df.agg({'price': 'mean'}).sort_values('price').round()

price
bedrooms           
1          317643.0
2          401373.0
0          409504.0
3          466232.0
11         520000.0
4          635420.0
33         640000.0
5          786600.0
10         819333.0
6          825521.0
9          894000.0
7          951185.0
8         1105077.0

3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [ ]:
grouped_df = df.groupby('bedrooms')
grouped_df.agg({'price': 'mean'}).sort_values('price').round()

3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [45]:
pd.crosstab(index=df['waterfront'],
            columns=df['view'])

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

3.4 Каких домов в зависимости от этажности и количества спален больше?

In [46]:
pd.crosstab(index=df['floors'],
            columns=df['bedrooms'])

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Больше всего 1-этажных домов с 3-мя спальнями (5455)

3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [54]:
df.pivot_table(index='condition',
               columns='grade',
               values='price',
               aggfunc='median',
               fill_value='-')

grade            1         3         4       5       6       7       8   \
condition                                                                 
1          142000.0         -  150000.0  190000  255000  403500  932500   
2                 -  280000.0  145000.0  180000  235000  305000  429000   
3                 -   75000.0  205000.0  234475  265000  357500  485000   
4                 -         -  238525.0  229975  295000  390000  571250   
5                 -  262000.0         -  227450  285475  456000  696000   

grade             9          10         11         12         13  
condition                                                         
1                  -          -          -          -          -  
2           715000.0  1752500.0          -          -          -  
3           689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4           823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          1078000.0  1650000.0  2050000.0  1990000.0          -